src/pages/posts/comparing-rust-smart-contract-storage-approaches.md

```diff
@@ -0,0 +1,515 @@
+
+## NEAR
+
+`near-sdk-rs` provides [collections](https://docs.rs/near-sdk/latest/near_sdk/collections/index.html) and, more recently, [store](https://docs.rs/near-sdk/latest/near_sdk/store/index.html) modules with efficient data structures to store large sets of data in the key-value store. Most code in the wild, including [the example fungible token example](https://github.com/near/near-sdk-rs/tree/master/examples/fungible-token), still use `collections`.
+
+```rust
+use near_sdk::collections::LookupMap;
+
+// this comes from the standards library:
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct FungibleToken {
+    /// AccountID -> Account balance.
+    pub accounts: LookupMap<AccountId, Balance>,
+}
+
+// then, in your contract, this is your main struct:
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Contract {
+    token: FungibleToken,
+}
+```
+
+Then, inside a method definition like `increment` above, you access those accounts using [LookupMap methods](https://docs.rs/near-sdk/latest/near_sdk/collections/struct.LookupMap.html). Some examples:
+
+```rust
+// add a new account to the map
+self.token.accounts.insert(&tmp_account_id, &0u128);
+
+// remove one
+self.token.accounts.remove(&tmp_account_id);
+
+// reduce someone's balance
+if let Some(new_balance) = balance.checked_sub(amount) {
+    self.token.accounts.insert(account_id, &new_balance);
+}
+```
+
+## Soroban
+
+Soroban also has [an reference Fungible Token implementation](https://github.com/stellar/soroban-examples/tree/main/token).
+
+`Identifier` is an account identifier, like a public key or account address.
+
+```rust
+use soroban_auth::Identifier;
+use soroban_sdk::contracttype;
+
+#[derive(Clone)]
+#[contracttype]
+pub enum DataKey {
+    Balance(Identifier),
+}
+```
+
+This is using [custom types](https://soroban.stellar.org/docs/examples/custom-types), which the simple example didn't need. The full reference implementation includes several other variants on this enum.
+
+Then, to use `DataKey`:
+
+```rust
+use soroban_sdk::{BigInt, Env};
+
+/// Get an account's balance, defaulting to zero
+pub fn read_balance(e: &Env, id: Identifier) -> BigInt {
+    let key = DataKey::Balance(id);
+    if let Some(balance) = e.data().get(key) {
+        balance.unwrap()
+    } else {
+        BigInt::zero(e)
+    }
+}
+
+/// Update an account's balance:
+fn write_balance(e: &Env, id: Identifier, amount: BigInt) {
+    let key = DataKey::Balance(id);
+    e.data().set(key, amount);
+}
+```
+
+As with the simple example, all data manipulations go through the [`e.data()`](https://docs.rs/soroban-sdk/0.2.1/soroban_sdk/data/struct.Data.html) interface.
+
+Like CosmWasm, and unlike NEAR, this avoids loading all data every time you need any data.
```

Comment:
I'm confused by this statement. The following works with NEAR. Also another aspect where riffs make sure you only load the data you need.

```rust
struct Empty;

#[near_bindgen]
impl Empty {
  // Example static method
  pub fn skips_loading_state() -> String {
    "empty".to_owned()
  } 
}

```

Reply:
This post does not mention Riffs, since Riffs has not yet been finalized / announced. So I'm comparing what is common in NEAR.

And yes, you can export functions that don't read any state. But as soon as you include `&self`, it deserializes all state, right?

Reply:
```suggestion
Like CosmWasm, and unlike NEAR, this avoids loading all contract state every time you need to access even just one field from your main contract struct.
```